File: src/main/java/itmo/is/project/model/module/production/Production.java
package itmo.is.project.model.module.production;

import itmo.is.project.model.Resource;
import jakarta.persistence.*;
import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Entity
@Table(name = "production")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class Production {

    @Embeddable
    public record CompositeKey (Integer blueprintId, Integer resourceId) {}

    @EmbeddedId
    private Production.CompositeKey id;

    @MapsId("blueprintId")
    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "blueprint_id", referencedColumnName = "id")
    private ProductionModuleBlueprint blueprint;

    @MapsId("resourceId")
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "resource_id", referencedColumnName = "id", nullable = false, updatable = false)
    private Resource resource;

    @NotNull
    @Min(1)
    @Column(name = "amount", nullable = false, updatable = false)
    private Integer amount;
}

File: src/main/kotlin/itmo/isproject/model/module/production/Production.kt
package itmo.isproject.model.module.production

import itmo.isproject.model.EmbeddedIdEntity
import itmo.isproject.model.resource.Resource
import itmo.isproject.model.resource.ResourceAmountHolder
import jakarta.persistence.*
import jakarta.validation.constraints.Min
import jakarta.validation.constraints.NotNull
import java.io.Serializable

@Entity
@Table(name = "production")
class Production(
    
    @MapsId("blueprintId")
    @OneToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "blueprint_id", referencedColumnName = "id")
    var blueprint: ProductionModuleBlueprint? = null,

    @MapsId("resourceId")
    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = "resource_id", referencedColumnName = "id", nullable = false, updatable = false)
    override var resource: Resource? = null,

    @NotNull
    @Min(1)
    @Column(name = "amount", nullable = false, updatable = false)
    override var amount: Int? = null

) : EmbeddedIdEntity<ProductionId>(), ResourceAmountHolder

@Embeddable
data class ProductionId(
    val blueprintId: Int?,
    val resourceId: Int?
) : Serializable